### ISSUE TITLE
pyinstaller app fails with 'ImportError' on backend modules.

### ISSUE THREAD

**Issue by jaraco** [jaraco/keyring#124]
I upgraded to Mac OS 10.9 and now when I build by app using pyinstaller, I'm getting 'ImportError: No module named backends.file'.  Using keyring 3.1 and python 2.7.5.

---
- Bitbucket: https://bitbucket.org/kang/python-keyring-lib/issue/124
- Originally reported by: Anonymous
- Originally created at: 2013-10-27T17:34:46.299

**Comment by jaraco** [comment]
OK, a little more information.  I upgraded a second system I have to 10.9 and my pyinstaller builds do work.  Difference is keyring 1.2 vs 3.1 and python 2.7.5 vs 2.7.3.

---

Original comment by: [mgoblue80](http://bitbucket.org/mgoblue80)

**Comment by jaraco** [comment]
I'm afraid I'm going to need a bit more information. In other systems, 3.1 is running just fine on Python 2.7.3 and 2.7.5. Here's where I would investigate:
- Could it be that the first system had a config file that specified a particular backend? Make sure that backend includes the full path to the desired backend. Try without the config file.
- Confirm that the module actually exists. Find where keyring is installed and look for 'keyring/backends/file.py'.

I was going to suggest other things to try like importing 'keyring.backends.file' manually, but I realize now there's no way to get at the backends without causing everything to be imported (as a side-effect of `import keyring`).

It's conceivable there's a race condition when importing a certain backend before another. However, I wouldn't expect that to happen using the default configuration (as the import behavior is very similar for all platforms) - so I'd still suspect the behavior is triggered by something in configuration.

Please let me know what you find.

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
Here's a little more information and answer to your questions:
- There was no config file on the old system.  I've never used the keyring config file feature.
- I did check and the module does exist.  I am using virtualenv and keying/backends/file.py exists.  I do get this message when I attempt to import manually:

```
#!python

$ python 
Python 2.7.5 (default, Aug 25 2013, 00:04:04) 
[GCC 4.2.1 Compatible Apple LLVM 5.0 (clang-500.0.68)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import keyring.backends.file
No handlers could be found for logger "keyring"
```

I get the same "No handlers..." message when I run my app without bundling with pyinstaller and the app seems to run fine.  

---

Original comment by: [mgoblue80](http://bitbucket.org/mgoblue80)

**Comment by jaraco** [comment]
Feel free to close this issue.  I've found an other solution to password management for my app.

---

Original comment by: [mgoblue80](http://bitbucket.org/mgoblue80)

**Comment by jaraco** [comment]
@mgoblue80  what other method you had find. Please explain, I am also facing issues with keyring

---

Original comment by: Hárísh Bárvékár

**Comment by jaraco** [comment]
Add a pyinstaller hook to expose backend modules. Ref #124

→ <<cset 8794f2aa67cd>>

---

Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
I understand better now. I see that the issue is due to the use of pyinstaller. I only realized this after having recently encountered a similar situation in py2exe.

I believe the issue is that keyring no longer eagerly imports all of the backends, so pyinstaller can't detect the use of the backends modules, and so fails to include it in the bundle.

According to [the docs](http://www.pyinstaller.org/export/develop/project/doc/Manual.html#listing-hidden-imports), it should be possible to create a hook for pyinstaller to work with these later versions of keyring.

In the aforementioned changeset, I've added a pyinstaller hook that _should_ correct the issue. Can someone try the hook according to the docs on using a private hook? If that works, I'll submit it upstream to pyinstaller.

---

Original comment by: Jason R. Coombs
